aten/src/ATen/native/LinearAlgebraUtils.h

```diff
@@ -347,4 +347,36 @@ static inline void checkUplo(const std::string& uplo) {
     "Expected UPLO argument to be 'L' or 'U', but got ", uplo);
 }
 
+static inline void checkSameDevice(const std::string& fn_name, Tensor result, Tensor input, const std::string& result_name = "result") {
+  TORCH_CHECK(
+      result.device() == input.device(),
+      fn_name,
+      ": Expected ", result_name, " and input tensors to be on the same device, but got ",
+      result_name, " on ", result.device(), " and input on ", input.device());
+}
+
+// Check the dtype of result and input tensors (for _out variants).
+// Most linear algebra functions have the same dtype for input and output
+// (either floating or complex type input), so we can check whether both result and input are compatible.
+// According to https://github.com/pytorch/pytorch/wiki/Developer-FAQ#how-does-out-work-in-pytorch
+// if input is complex result can only be complex, if input is float result can be both float and complex
+static inline void checkLinalgCompatibleDtype(const std::string& fn_name, Tensor result, Tensor input, const std::string& result_name = "result") {
+  bool can_cast = c10::canCast(input.scalar_type(), result.scalar_type());
+  TORCH_CHECK(
+      can_cast,
+      fn_name,
+      ": Expected ", result_name, " to be compatible with ", input.scalar_type(), " dtype, but got ",
```

Comment:
Expected input's dtype to be (safely) castable to result's dtype.

We don't have a notion of "compatible" dtypes.